File: runtime/go/document_examples_test.go
package runtime

import (
	"encoding/json"
	"os"
	"testing"
)

// Both validators consume the document's actual bytes and visible bindings.
// Unavailable entries remain in the table, with a reason and no invented value.
func TestDocumentExamplesConformance(t *testing.T) {
	data, err := os.ReadFile("../../conformance/tables/examples.json")
	if err != nil {
		t.Fatal(err)
	}
	var table []struct {
		Name    string
		Schemas map[string]json.RawMessage
		Rows    []struct {
			Family, Path, To, Member string
			Expression, Value        json.RawMessage
			Bindings                 map[string]struct{ Type, Family string }
			Unavailable              *struct{ Kind, Reason string }
		}
	}
	if err := json.Unmarshal(data, &table); err != nil {
		t.Fatal(err)
	}
	proof := 0
	for _, corpus := range table {
		schemas := map[string]*Schema{}
		for name, wire := range corpus.Schemas {
			schemas[name] = MustSchema(string(wire), schemas)
		}
		for _, row := range corpus.Rows {
			t.Run(corpus.Name+"/"+row.Family+"/"+row.Path, func(t *testing.T) {
				if row.Unavailable != nil {
					if len(row.Value) != 0 || row.Unavailable.Reason == "" || row.Unavailable.Kind != "limit" && row.Unavailable.Kind != "impossible" {
						t.Fatalf("invalid unavailability: %+v", row)
					}
					return
				}
				if !json.Valid(row.Value) {
					t.Fatal("missing example value")
				}
				if row.Family == "proof" {
					proof++
				}
				types, families := map[string]any{}, map[string]*Schema{}
				for name, binding := range row.Bindings {
					if binding.Family != "" {
						families[name] = schemas[binding.Family]
						if families[name] == nil {
							t.Fatalf("missing family binding %s", binding.Family)
						}
					} else {
						types[name] = binding.Type
					}
				}
				value := row.Value
				if row.To != "" {
					frame, err := decodeFrame(value)
					if err != nil {
						t.Fatal(err)
					}
					local, remote := "s:", "c:"
					if row.To == "client" {
						local, remote = "c:", "s:"
					}
					prefix := remote
					if frame.Kind == "response" {
						prefix = local
					}
					if frame.ID != "" && !validID(frame.ID, prefix) {
						t.Fatal("invalid frame id")
					}
					if row.Member != "" {
						var members map[string]json.RawMessage
						if err := json.Unmarshal(value, &members); err != nil {
							t.Fatal(err)
						}
						value = members[row.Member]
					}
				}
				if len(row.Expression) != 0 {
					if err := schemas[row.Family].Bind(types, families).ValidateExpressionRaw(MustTypeExpression(string(row.Expression)), value); err != nil {
						t.Fatalf("%s: %v", value, err)
					}
				}
			})
		}
	}
	if proof < 30 {
		t.Fatalf("only %d proof examples are held", proof)
	}
}

File: runtime/ts/src/document-examples.test.ts
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { test } from 'node:test';
import { decodeEnvelope } from './envelope.ts';
import { createValidator, type Validator, type WireFamily, type TypeExpression, type Slots } from './validate.ts';

test('every concrete document example validates with its visible bindings', () => {
  const table = JSON.parse(
    readFileSync(new URL('../../../conformance/tables/examples.json', import.meta.url), 'utf8'),
  ) as {
    schemas: Record<string, Record<string, WireFamily>>;
    rows: {
      corpus: string;
      family: string;
      path: string;
      expression?: TypeExpression;
      value?: unknown;
      bindings?: Record<string, { Type?: string; Family?: string }>;
      unavailable?: { Kind: string; Reason: string };
      to?: string;
      member?: string;
    }[];
  };
  let proof = 0;
  const checkouts: Record<string, Record<string, Validator>> = {};
  for (const [corpus, schemas] of Object.entries(table.schemas)) {
    const imported: Record<string, Validator> = {};
    for (const [name, wire] of Object.entries(schemas)) imported[name] = createValidator(wire, imported);
    checkouts[corpus] = imported;
  }
  for (const row of table.rows) {
    const imported = checkouts[row.corpus]!;
    const where = `${row.corpus}/${row.family}/${row.path}`;
    if (row.unavailable) {
      assert(!Object.hasOwn(row, 'value'), where);
      assert(row.unavailable.Reason, where);
      assert(['limit', 'impossible'].includes(row.unavailable.Kind), where);
      continue;
    }
    assert(Object.hasOwn(row, 'value'), where);
    if (row.family === 'proof') proof++;
    const validate = imported[row.family]!;
    const slots: Slots = Object.fromEntries(
      Object.entries(row.bindings ?? {}).map(([name, binding]) => {
        if (binding.Family) {
          assert(imported[binding.Family], where);
          return [name, { name: binding.Family, validate: imported[binding.Family]! }];
        }
        return [name, { type: binding.Type as TypeExpression, validate }];
      }),
    );
    assert.doesNotThrow(() => {
      let value = row.value;
      if (row.to) {
        const [local, remote] = row.to === 'client' ? ['c:', 's:'] : ['s:', 'c:'];
        decodeEnvelope(JSON.stringify(value), local, remote);
        if (row.member) value = (value as Record<string, unknown>)[row.member];
      }
      if (row.expression !== undefined) validate(row.expression, value, '$', slots);
    }, where);
  }
  assert(proof >= 30, `only ${proof} proof examples are held`);
});
